clients/drcachesim/tracer/tracer.cpp

```diff
@@ -921,34 +1027,56 @@ process_buffer_for_physaddr(void *drcontext, per_thread_t *data, size_t header_s
+            // of markers on which we waste buffer output overhead.
+            // XXX: We could count them up and do a memmove if the count is small
+            // and we have space in the redzone?
+            if (!emitted) {
+                // We need to be sure to emit the initial thread header if this is before
+                // the first regular buffer and skip it in the regular buffer.
+                if (header_size > buf_hdr_slots_size) {
+                    size_t size = reinterpret_cast<offline_instru_t *>(instru)
+                                      ->append_thread_header(data->v2p_buf,
+                                                             dr_get_thread_id(drcontext),
+                                                             get_file_type());
+                    ASSERT(size == data->init_header_size, "inconsistent header");
+                    skip = data->init_header_size;
+                    v2p_ptr += size;
+                    header_size += size;
+                }
+                v2p_ptr += add_buffer_header(drcontext, data, v2p_ptr);
+                emitted = true;
+            }
+            if (v2p_ptr + 2 * instru->sizeof_entry() - data->v2p_buf >=
+                static_cast<ssize_t>(get_v2p_buffer_size())) {
+                NOTIFY(1, "Reached v2p buffer limit: emitting multiple times\n");
+                data->num_phys_markers +=
+                    output_buffer(drcontext, data, data->v2p_buf, v2p_ptr, header_size);
+                v2p_ptr = data->v2p_buf;
+                v2p_ptr += add_buffer_header(drcontext, data, v2p_ptr);
+            }
+            if (success) {
+                v2p_ptr += instru->append_marker(
+                    v2p_ptr, TRACE_MARKER_TYPE_PHYSICAL_ADDRESS, phys);
+                v2p_ptr += instru->append_marker(v2p_ptr,
+                                                 TRACE_MARKER_TYPE_VIRTUAL_ADDRESS, virt);
+            } else {
+                // For translation failure, we insert a distinct marker type, so analyzers
+                // know for sure and don't have to infer based on a missing marker.
+                v2p_ptr += instru->append_marker(
+                    v2p_ptr, TRACE_MARKER_TYPE_PHYSICAL_ADDRESS_NOT_AVAILABLE, phys);
             }
-            memmove(mem_ref + instru->sizeof_entry(), mem_ref, buf_ptr - mem_ref);
-            // For translation failure, we insert a distinct marker type, so analyzers
-            // know for sure and don't have to infer based on a missing marker.
-            mem_ref += instru->append_marker(
-                mem_ref,
-                success ? TRACE_MARKER_TYPE_PHYSICAL_ADDRESS
-                        : TRACE_MARKER_TYPE_PHYSICAL_ADDRESS_NOT_AVAILABLE,
-                phys);
-            buf_ptr += instru->sizeof_entry();
-            ++data->num_phys_markers;
         } else {
             // For online we replace the virtual with physical.
```

Comment:
Not sure if this is documented on our main docs.

Reply:
Maybe add details to https://dynamorio.org/sec_drcachesim_phys.html?

Reply:
I plan to change online to also use the markers, and change the dr$sim simulators to use the markers as well, and update that very docs section at that time.  A forthcoming PR.  I added this docs page to the existing XXX comment here.